kmall.py

```diff
@@ -3425,37 +3481,201 @@ def seek_next_startbyte(self, file_length, start_ptr=0):
                     return True
 
-    def sequential_read_records(self, first_installation_rec=False, start_ptr=0, end_ptr=0):
-        ########### STILL IN DEVELOPMENT ############
-        raise NotImplementedError('sequential_read_records is still in development')
-
-        # data we are reading
-        # systemID is not the same thing as serial number, only seems to be in the IIP
-        recs_categories = {'SKM': ['header.dgtime', 'sample.KMdefault.roll_deg', 'sample.KMdefault.pitch_deg',
-                                   'sample.KMdefault.heave_m', 'sample.KMdefault.heading_deg',
-                                   'sample.KMdefault.latitude_deg', 'sample.KMdefault.longitude_deg',
-                                   'sample.KMdefault.ellipsoidHeight_m'],
-                           'IIP': ['header.dgtime', 'install_txt'],
-                           'MRZ': ['header.dgtime', 'cmnPart.pingCnt', 'pingInfo.soundSpeedAtTxDepth_mPerSec',
-                                   'pingInfo.numTxSectors', 'header.systemID', 'txSectorInfo.txSectorNumb',
-                                   'txSectorInfo.tiltAngleReTx_deg', 'txSectorInfo.sectorTransmitDelay_sec',
-                                   'txSectorInfo.centreFreq_Hz', 'sounding.beamAngleReRx_deg',
-                                   'sounding.txSectorNumb', 'sounding.detectionType', 'sounding.qualityFactor', 
-                                   'sounding.twoWayTravelTime_sec', 'pingInfo.modeAndStabilisation',
-                                   'pingInfo.pulseForm', 'pingInfo.depthMode'],
-                           'SVP': ['time_sec', 'sensorData.depth_m', 'sensorData.soundVelocity_mPerSec']}
-        
+    def _divide_rec(self, rec):
+        """
+        MRZ comes in from sequential read by time/ping.  Each ping may have multiple sectors to it which we want
+        to treat as separate pings.  Do this by generating a new record for each sector in the ping.
+        """
+        if self.datagram_ident != 'MRZ':
+            return [rec]
+        elif rec['pingInfo']['numTxSectors'] == 1:
+            return [rec]
+        else:
+            totalrecs = []
+            pingtime = rec['header']['dgtime']
+            for sec in rec['txSectorInfo']['txSectorNumb']:
+                split_rec = copy.copy(rec)
+                split_rec['txSectorInfo'] = {k: v[sec] for (k,v) in rec['txSectorInfo'].items()}
+                rx_index = np.where(np.array(rec['sounding']['txSectorNumb']) == sec)
+                split_rec['sounding'] = {k: np.array(v)[rx_index] for (k,v) in rec['sounding'].items()}
+
+                # ping time equals datagram time plus sector transmit delay
+                split_rec['header']['dgtime'] = pingtime + split_rec['txSectorInfo']['sectorTransmitDelay_sec']
+
+                totalrecs.append(split_rec)
+            return totalrecs
+
+    def _pad_to_dense(self, arr, padval=999.0, maxlen=500, override_type=None, detectioninfo=False):
```

Comment:
Maybe add in the doc string why one would want to "pad to dense". If one doesn't know what you're up to, it be hard to understand the purpose of the function. [I think it's because you need a rectangular array of records for down-stream processing, but I'm not 100% sure myself.]

Reply:
done